### PR TITLE
Add type checks for `Contains`, `NotContains`, `Subset` and `NotSubset`

### PR DESCRIPTION
**Pull request by g7r**
The code like `assert.NotContains(t, list, element)` is very likely to be
broken when type of element or type of list element changes. This change
makes testify more paranoid forcing mentioned code to break when types
doesn't match.

### CODE REVIEW

**Comment by devdinu** [comment]
@g7r could you share test cases for the current `Contains` where you expect it to fail, but it behaves differently. helps to understand the need and review the fix. Also i'm curious if its required why it can't be part of `includeElement`?

Also you could add tests against `Contains, Subset, NotSubset` and not private function, as it has side effect on those functions. and easy to verify.

**Comment by g7r** [comment]
The case is from real life of course. Let's say we have a collection with documents. To access the collection we have DAO. Document identifier is a plain string:
```
type DAO interface {
    FindDocumentIDs(q Query) ([]string, error)
}
```
Some test cases are written for DAO:
```
func TestDAO(t *testing.T) {
    var dao DAO = ...
    documentIDs, err := dao.FindDocumentIDs(...)
    require.NoError(t, err)
    require.NotContains(t, documentIDs, "DocumentFoo")
}
```
After some time have passed, we have a lot of entities with their identifiers specified with `string` too. To make code semantic more explicit we introduce named types for each entity identifier type. For example:
```
type DocumentID string
```
And voila, the test is completely broken. The compiler and test harness don't see the problem. Even if `documentIDs` contains "DocumentFoo" the test will be green.
Thats the kind of problem I'm trying to solve with this PR.

As for embedding `assertIncludesElementTypes` into `includesElement`, that was the first version of my changeset. But it deemed more difficult to report clear error messages in that case. Third `bool` result for `includesElement` would be too much.

I'll add tests for fixed cases into `Contains`, `NotContains`, `Subset` and `NotSubset` a bit later if everything aside of that is OK in this PR.

**Comment by devdinu** [comment]
Yep, `NotContains(t, []DocumentID{"id"}, "id")` pass without failure. Having a separate function is fine. can you change the way you pass hardcoded values to  assertIncludeElementTypes `elementIsList` rather infer it.

File: assert/assertions.go [hunk 1]
```diff
@@ -598,6 +598,56 @@ func NotEqual(t TestingT, expected, actual interface{}, msgAndArgs ...interface{
 
 }
 
+// assertIncludeElementTypes checks whether types are compatible as incompatible types indicate programmer error.
+// return true if list and element types are compatible and includeElement can possibly return found=true.
+// return false if list and element types aren't compatible and includeElement will always return found=false.
+func assertIncludeElementTypes(list interface{}, element interface{}, elementIsList bool) bool {
+
+	var elementType reflect.Type
+	if element != nil {
```
[the remark this answers is not in the document]
Comment (g7r) [reply assert/assertions.go]: We can't blindly "unwrap" `element` if it is of kind `reflect.Array`, because it'll break such constructs as `Contains(t, [...][1]int{{1}, {2}, {3}}, [1]int{1})`. "Unwrapping" `reflect.Slice` is also dangerous: `Contains(t, [][]int{{1}, {2}, {3}, nil}, []int(nil))`. It is possible to unwrap `element` with respect to `list` type. But it'll increase complexity of the code without much profit.
Another problem is that it will break things like `Subset(t, []int{1, 2}, int(0))`. Impossible type combination but `assertIncludeElementTypes` will be happy.
Here is implementation of `assertIncludeElementTypes` without `elementIsList`, for the reference:
```
func assertIncludeElementTypes(list interface{}, element interface{}) bool {
	var listElementType reflect.Type
	switch listType := reflect.TypeOf(list); listType.Kind() {
	case reflect.String:
		listElementType = listType

	case reflect.Map:
		listElementType = listType.Key()

	case reflect.Slice, reflect.Array:
		listElementType = listType.Elem()

	default:
		return false
	}

	if element == nil {
		switch listElementType.Kind() {
		case reflect.Ptr, reflect.Slice, reflect.Map, reflect.Interface, reflect.Func:
			return true
		default:
			return false
		}
	}

	isCompatible := func(t reflect.Type) bool {
		if t.Kind() == reflect.Interface {
			// Anything can be hidden behind interface
			return true
		}

		return t.AssignableTo(listElementType)
	}

	var elementType reflect.Type
	elementType = reflect.TypeOf(element)
	if !isCompatible(elementType) {
		switch elementType.Kind() {
		case reflect.Slice, reflect.Array:
			if !isCompatible(elementType.Elem()) {
				return false
			}

			elementType = elementType.Elem()

		default:
			return false
		}
	}

	return true
}
```
It is broken in ways I've described above.

File: assert/assertions.go [hunk 2]
```diff
@@ -598,6 +598,56 @@ func NotEqual(t TestingT, expected, actual interface{}, msgAndArgs ...interface{
...
+			switch elementType.Kind() {
+			case reflect.Array, reflect.Slice:
+				elementType = elementType.Elem()
+			default:
+				return false
+			}
+		}
+	}
+
+	listType := reflect.TypeOf(list)
```
[the remark this answers is not in the document]
Comment (g7r) [reply assert/assertions.go]: Agreed, will update the code. As a side effect it'll make `Contains` a bit more strict. The cases like that will become illegal:
```go
type T string
Contains(t, "123", T("1"))
```

**Comment by g7r** [comment]
@devdinu, I've pushed heavily updated code.
    1. Splitted `assertIncludeElementTypes` to two functions for clarification.
    2. Removed redundant type checks in `Subset` and `NotSubset`.
    3. Removed now-unnecessary panic handler in `includeElement`.
    4. Moved `subset == nil` lower in `Subset` for type checks to happen.
    5. Added test cases covering new type checks into tests that indirectly use them: `TestContains`, `TestNotContains`, `TestSubset` and `TestNotSubset`.

**Comment by g7r** [comment]
@devdinu, Is something wrong with this PR?

**Comment by devdinu** [comment]
@g7r PR looks good to me. @ernesto-jimenez could review and merge it.

**Comment by g7r** [comment]
@ernesto-jimenez can I be of any help with this PR?

**Comment by ernesto-jimenez** [comment]
@g7r my main question is whether this changes behaviour when using interface types. If so, how?

Interface types usually the trickiest when changing behaviours to make assertions more strict with types.

**Comment by g7r** [comment]
@ernesto-jimenez 
In `Contains` and `NotContains`:
```
var list []SomeInterface = ...
var value = ...
assert.Contains(t, list, value)
assert.NotContains(t, list, value)
```
I will assert that `value` type implements `SomeInterface`.

In `Subset` and `NotSubset` (1):
```
var list []SomeInterface = ...
var subset []AnotherInterface = ...
assert.Subset(t, list, subset)
assert.NotSubset(t, list, subset)
```
I will not assert anything.

In `Subset` and `NotSubset` (2):
```
var list []SomeInterface = ...
var subset []NonInterfaceType = ...
assert.Subset(t, list, subset)
assert.NotSubset(t, list, subset)
```
I will assert that `NonInterfaceType` implements `SomeInterface`.

**Comment by ernesto-jimenez** [comment]
@g7r great, could you add those to the unit tests so the behaviour is clear and nobody accidentally breaks it in the future?

Thanks!

**Comment by g7r** [comment]
@ernesto-jimenez 
These cases are already tested. The thing is that it is somewhat difficult to thoroughly test all type compatibility cases using only public methods from package `assert`. So tests for `Contains`, `NotContains`, `Subset` and `NotSubset` include only basic type compatibility cases. Only to assert that type compatibility checks are not broken at all. And there are additional tests that test type compatibility check internals which are quite exhaustive (all aforementioned cases are included).

**Comment by ernesto-jimenez** [comment]
We should also test some these thorny cases through the public API so it's clear those are cases we support and must keep supported.

Otherwise, somebody might refactor/change the internals in the future, break some of those cases, and go unnoticed.

**Comment by g7r** [comment]
OK, I'll add more cases to tests for `NotContains` and `NotSubset` as those are mostly vulnerable to the issue.

**Comment by g7r** [comment]
Adding tests for `Contains` and `Subset` is pointless because there is no way to distinguish negative cases.
